pkg/schema/typesystem.go

```diff
@@ -64,3 +67,127 @@ func (ts *TypeSystem) getDefinition(ctx context.Context, definition string) (*De
+		// TODO(jschorr): Change this once we load schema as a single unit.
+		return Traits{
+			AllowsCaveats:    true,
+			AllowsExpiration: true,
+		}, nil
+	}
+
+	vts, err := ts.GetValidatedDefinition(ctx, filter.OptionalResourceType)
+	if err != nil {
+		// If the type doesn't exist, we return a KnownEmptyFilterError.
+		if errors.As(err, &DefinitionNotFoundError{}) {
+			return Traits{}, NewKnownEmptyFilterErr(fmt.Sprintf("resource type '%s' does not exist in schema", filter.OptionalResourceType))
+		}
+
+		return Traits{}, err
+	}
+
+	// If there are no relations defined in the filter, return the traits for all relations.
+	if filter.OptionalResourceRelation == "" {
+		traits, err := vts.PossibleTraitsForAnyRelation()
+		return traits, err
+	}
+
+	// Check if the relation exists.
+	if _, ok := vts.GetRelation(filter.OptionalResourceRelation); !ok {
+		// If the relation doesn't exist, we return a KnownEmptyFilterError.
+		return Traits{}, NewKnownEmptyFilterErr(fmt.Sprintf("relation '%s' does not exist on resource type '%s'", filter.OptionalResourceRelation, filter.OptionalResourceType))
+	}
+
+	// If there is not subject filter, use the relation only.
+	if len(filter.OptionalSubjectsSelectors) == 0 {
+		traits, err := vts.PossibleTraitsForAnySubject(filter.OptionalResourceRelation)
+		return traits, err
+	}
+
+	// Otherwise, build a combined traits set for all the subject filters.
+	var combinedTraits Traits
+	for _, subjectFilter := range filter.OptionalSubjectsSelectors {
+		// If the subject type is not specified, we need to get the traits for any subject.
+		if subjectFilter.OptionalSubjectType == "" {
+			traits, err := vts.PossibleTraitsForAnySubject(filter.OptionalResourceRelation)
+			return traits, err
+		}
+
+		// Otherwise, get the traits for the specific subject type and combine them.
+		traits, err := vts.PossibleTraitsForSubject(filter.OptionalResourceRelation, subjectFilter.OptionalSubjectType)
+		if err != nil {
+			// If the subject type doesn't exist on this relation, fall back to traits for any subject
+			// to err on the side of caution.
+			traits, err = vts.PossibleTraitsForAnySubject(filter.OptionalResourceRelation)
+			if err != nil {
+				return Traits{}, err
+			}
+		}
+
+		combinedTraits = combinedTraits.union(traits)
+	}
+
+	return combinedTraits, nil
+}
+
+// applyExpirationFilter adjusts traits based on the expiration filter option
+func applyExpirationFilter(traits Traits, expirationOption datastore.ExpirationFilterOption) (Traits, error) {
+	switch expirationOption {
+	case datastore.ExpirationFilterOptionNone:
+		// No filtering, return traits as-is
+		return traits, nil
+
+	case datastore.ExpirationFilterOptionHasExpiration:
+		// Only relationships WITH expiration are allowed
+		if !traits.AllowsExpiration {
+			// If the relation doesn't support expiration, the filter is empty
+			return Traits{}, NewKnownEmptyFilterErr("filter requires expiration but the schema relation(s) do not support expiration")
```

Comment:
```suggestion
			return Traits{}, NewKnownEmptyFilterErr("the filter requested relationships with expiration but the filter relation does not support expiration")
```

also why the plural (s)? you can't specify more than one relation filter right?

Reply:
Changed